### ISSUE TITLE
Regex engines and grammars in the wild

### ISSUE THREAD

**Issue by dumblob** [dumblob/tmp#1]
Continuation of the discussion under https://github.com/vlang/v/commit/8ea0c08a388044c1f646798e919be811fc83cd48 .

The points I'd like to discuss at the beginning are these:

1. I claim, that Python regexes (`re` module) are incoherent due to their [varying greediness and non-greediness rules](https://stackoverflow.com/a/33907811 ) (which are quite context sensitive which gets really complex/unreadable very early) compared to POSIX EREs (which are always maximally greedy, no matter what).

2. I'm not aware of any incoherence in [POSIX EREs](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_04 ) (no matter whether regarding greediness or any other feature). Anything I'm missing?

I'd be glad to hear your opinions and experience. Thanks!

**Comment by penguindark** [comment]
Hi, thanks for creating this discussion space :)
Let's start our discussion, and thanks again for this beautiful confrontation.

That python/perl regex are incoherent you are right 100% but they are more usable in every day problems then the posix standard, Posix ERE for sure is 100% formal and correct (it has a fixed and well definited grammar after all!)

Before continue, please, let me say that I'm totally agnostic about regex standards and implementations, I see regex as a simple tool no more no less. ;)
In fact some Posix ERE features will be ported in the v regex engine that I'm writing.

Using regex almost every day from about 1997 and having used both of the standards, IMHO the perl dialect is the best for "normal use", let me explain.
 
What I mean for incohernece in the PERE (Posix ERE) are some choice that they made in the standard that not helped PERE to become the "stadard de facto" that as you know at present time is an "hybrid" PERL regex defintion, with all its problems and formal defects.
To put it roughly PERE is too much academic and far from the world of work.

I agree that complicate regex in python/perl are a mess, but if you need a big regexp expression probability you are using the wrong tool :)

If i remember well, about ten years ago, a Project Manager (that was working with me)  was in love with regex and was forcing it everywhere, i simply shown him that ten line of C code solved the problem that 2Kbyte of regex wasn't able to solve.
Regex are powerfull tools but force their use cause more troubles and slow solutions.
And, always, in my humble opinion and experience, if the regex is more than a line, other options are probably the best choice for your problem.

If we speak about formality and correctness PERE is the best, but on "human" side it is prolix and very long to write and has very low comunity support.
Usually I work on Gigabytes or Terabytes of text data with regex and in a practical usage I prefer python/perl dialect, but this only my personal opinion.

Now please  tell me your experience and why you think that PERE is the best choice :)
Tnx for your time.

**Comment by dumblob** [comment]
That's interesting - I have the very same opinion on everything you wrote, so there is not much to discuss and it's probably more about subjective needs :wink:.

There is one thing I value more (because the work I'm doing) than the fact, that PEREs are prolix. Namely the multiplatformity and actually speed. For PEREs there are multiple extremely fast engines (look e.g. at different AWKs). I use PEREs on Windows (most often through [git for windows]() as that's what most companies & clients have already installed), on macOS and all unix systems. I use them for gigabytes of data (I can't remember whet it was terabytes last time - maybe never) without any hiccups. On the other hand, for inputs bigger than very few megabytes, PEGs perform basically always better (because they don't have to read the whole input).

So, in other words, I'm trading slightly (yeah, from my experience it's just about twice to three times more concise to write the same PERE using PCRE) more verbose syntax for full multiplatformity, easier to read regexes (that's subjective, but because I work in several languages like POSIX stuff, JavaScript, Python, sometimes Perl and sometimes Lua, I absolutely don't remember what are the differences between `\w` in all of them, so I learned to avoid anything like that for clarity and **guarantee**) and usually greater speed.

Generally if it were possible to change history, I'd only do just one enhancement to PEREs - namely added lookahead support (and if easy to implement also lookbehind support, but not necessarily). This would cut down the length of many regexes I write to maybe about half, in which case I'd get the same "visual performance" to PCREs (visual performance metric: length of the regex needed to describe the problem without using any advanced stuff like `\w` because of varying definitions between languages and environments), but with guarantees (both correctness and speed), full multiplatformity, speed tendency (though lookahead is by definition not performant and highly prone to writing extremely slow regexes) and more or less easy implementation.

Notice, I'm comparing mostly to PCREs as that's what I prefer to Python regexes, but it shall be valid also for other regex engines.

What are your thoughts?

**Comment by penguindark** [comment]
I understand well your point of view, and the portability is a big problem.

The v regex was born as a tiny module with a minimal support, when I started to write it I chose to take inspiration from  the syntax of PCRE for two different reason:

- It is well know and the most common one
- For short regex query is more quick to write them

And I thought that in this moment of the development of V we need some "marketing" trick ;) to help the language to grow, an have a minimum PCRE support seems a good choice.

I also use and used a lot of languages in my programming life.
I started write code in 1982, I know, I'm old :(  and regex interoperability, in my experince, is near impossible to achive, but you are right, if it possible to achive it we can renunce other features.

The regex module probably will have few other steps of development and I hope that other persons will help me in this task.
But for more "heavy" use I'm sure that have third party module that link C libraries like pcre2 or re2 can be the good solution.
Think the v regex module as a very small simple regex parser with not full compliance but usefull for a quick and dirty use in small project.

Like the python re module vs regex module, just to have an example.

Have you ever tried to write a regex engine in your spare time?
If you decide to write one and have the constraint to stay under the 2k lines of code in v, what will be your choices? =)
I'm curious to have another point of view on a possible development.

**Comment by dumblob** [comment]
>Have you ever tried to write a regex engine in your spare time?

No, just many different parsers, interpreters, etc. You're few steps ahead with your effort :wink:.

>If you decide to write one and have the constraint to stay under the 2k lines of code in v, what will be your choices? =)

I think I'd try hard to implement **full** POSIX ERE compatibility **except for** both collating sequences (`[.ch.]`) and character equivalency (`[=e=]`). Then I'd implement support for both UTF-8 and byte-wise characters. And if I had some lines left to 2k, I'd implement lookahead syntax. And then if I still had some lines left, I'd implement "lazy" modifiers like `?` in `{m,n}?` and maybe other places (similar to Python `re`).

What about your ideals? I'm asking because I assume, that in your *V* regex implementation not all your visions came true.

**Comment by penguindark** [comment]
You are right, my ideal implementation of regex is not what I wrote for V.

> I think I'd try hard to implement full POSIX ERE compatibility except for ...

We all want the standard, but with "some" modifications. :wink: 

I started thinking about a DFA, but after some test I saw the code become very complicated and that some features were near impossible to achieve without tricks.
Than I changed to NFA but I wanted a small memory footprint and I chose to do not use recursion,
with recursion the code would have been much smaller and probably more readable.

After that I tried to write the most clear and didactic code and avoid explosion of features.
One thing that is not changed from my basic idea is that the "atomic" base of the regex engine is the "token" not the char, this allowed me to write the majority of functionalities in a very simple way.

By the way just yesterday I added the first implementation for:
- {m,n}? no greedy modifier
- byte-wise characters management
:wink: 

For now I think that I'll stop the development of other features and start work on speed up the find function.
Unfortunately my spare time, in this period, is very poor. :sob: 
Any suggestion on the "find" speed up methodology? :smiley:

**Comment by dumblob** [comment]
>By the way just yesterday I added the first implementation for: ...

Yes, I've already reacted there and apparently you also already discovered my thoughts there :wink:.

>Unfortunately my spare time, in this period, is very poor. 

Same on my side.

>Any suggestion on the "find" speed up methodology?

I'll take a look at the code. Generally nowadays the speed is not any more about "less branching means faster code", but about "never communicate with memory" (i.e. first load everything to registers and caches, then never store until you're completely done with the data and then don't touch the data any more - you can imagine that as windowing with couple of other constraints). We might also consider adding padding to buffers to allow manual loop unrolling to e.g. counts of 8 to allow the compiler to again merge these unrolled consecutive computations into SIMD (or other vector) instructions - see also https://github.com/vlang/v/issues/1839 .

**Comment by penguindark** [comment]
Interesting ideas, I'll take a look to find papers about your proposal, let see what I can achieve :)

**Comment by dumblob** [comment]
Regarding loop unrolling, the basic approach would look like follows.
```v
/* in get_charb() */
	for tmp < char_len {
		ch = (ch << 8) | in_txt[i+tmp]
		tmp++
	}
```
would become
```v
	a := u32( in_txt[i] )
	b := if char_len > 1 { u32( in_txt[i+1] ) << 8 } else { u32( 0 ) }
	c := if char_len > 2 { u32( in_txt[i+2] ) << 16 } else { u32( 0 ) }
	d := if char_len > 3 { u32( in_txt[i+3] ) << 24 } else { u32( 0 ) }
	return a | b | c | d, char_len
```
or maybe (this shows a general approach quite well - compute something very quickly and fully in parallel and then throw away parts which are not useful; it shows at the same time yet another extremely useful principle in optimization - lookup tables; on the other hand this is the only example which requires special property of the input data - in this case the padding of 3 bytes :cry:)
```v
	// this assumes in_txt is padded
	a := u32( in_txt[i] | in_txt[i+1] << 8 | in_txt[i+2] << 16 | in_txt[i+3] << 24 )
	x := [0x000000ff, 0x0000ffff, 0x00ffffff, 0xffffffff]
	return a & x[char_len], char_len
```
Note, I removed the dependencies between expressions/computations which might allocate few registers more in case of a really dumb compiler (which actually wouldn't allocate registers, but only stack space anyway), but in 99.9% cases rather allows the compiler to discover, that `a` `b` `c` `d` are all independently computed, have the same size (I think the compiler would have it more difficult if e.g. `a` was just `byte` instead of the same size as the SIMD instruction output) and thus can be squashed into just one SIMD instruction because the "inputs" (i.e. `i` and `ch`) don't change throughout the sequence and the sequence does basically the same operation on each line. On the other hand I'm not sure whether just these lines standalone would the compiler discover and maybe I'd try to "nudge" the compiler by enclosing this sequence (except for the last line) into a one-iteration loop, because compilers usually try to optimize only [tight **loops**](https://en.wiktionary.org/wiki/tight_loop ). But when doing such manual optimization, always look at the generated assembly code instead of guessing (let *V* compiler to output the generated C code and then run e.g. `gcc -S -o myassembly.s`).

Or in this case maybe some casting (beware of big/little endian - not sure how this works in *V*) could generate yet slightly faster code.
```v
	ch = u32( in_txt[i..i+char_len] ) >> 4 - char_len
```

In this particular case the following would probably be the fastest (should result in zero or one jump and one vector instruction), but most verbose :cry:
```v
	match char_len {
		1 { return u32( in_txt[i] ), char_len }
		2 { return u32( in_txt[i] ) | (u32( in_txt[i+1] ) << 8), char_len }
		3 { return u32( in_txt[i] ) | (u32( in_txt[i+1] ) << 8) | (u32( in_txt[i+2] ) << 16), char_len }
		else { return u32( in_txt[i] ) | (u32( in_txt[i+1] ) << 8) | (u32( in_txt[i+2] ) << 16 ) | (u32( in_txt[i+3] ) << 24), char_len }
	}
```

I hope this showed few cases of non-algorithmic optimization :wink:. Note also, that I changed the semantics what the code does, because it didn't look correct to me (the original loop always skips the first byte from my understanding, but that doesn't sound right to me - so bear with me if I overlooked something :wink:).

**Comment by dumblob** [comment]
Didn't check generated assembly, but I assume that changing all the `for tmp >= 0 {
` loops in `get_char_class()` to C-like for loops (which *V* should support and which should get translated to the exact C equivalent) would increase the probability, that the compiler will notice, that it's a plain counted loop (not a general loop) and that there are no dependencies between iterations and thus unroll the whole loop and in a later stage vectorize it. Maybe current compilers are very smart and will always discover it even if it's not the standard counted loop, but to increase the probability, I'd do change it :wink:.

**Comment by dumblob** [comment]
In `parse_char_class()` I think all references to `BSLS_VALIDATOR_ARRAY[c]` could be changed to `x` - is it so?

**Comment by dumblob** [comment]
```v
	for pc1 < pc {
		if re.prog[pc1].ist == IST_DOT_CHAR {
			tmp_count++
		}
		pc1++
	}
```
could be optimized by matching e.g. 4 or better 8 characters conditions in one iteration by finding out how many whole iterations can be made in `pc` (i.e. using the trick `pc & (-8)` - beware of sign of `pc` though), then iterating this number divided by 8 (division again using bit shift) and then having an epilogue for the rest zero, one, ... up to 7 remaining computations (this epilogue must not be that fast any more, so it might be implemented like your current loop by looping over one computation).

Changing the syntax of the for loop head to C-like syntax could maybe also "nudge" the compiler a bit more.

```v
	// not sure whether this multi-assignment syntax made it already to V
	for a, b, c, d, e, f, g, h := 0, 0, 0, 0, 0, 0, 0, 0; pc1 < pc; pc1+=8, tmp_count += a + b + c + d + e + f + g + h {
		a := if re.prog[pc1].ist == IST_DOT_CHAR { 1 } else { 0 }
		b := if re.prog[pc1+1].ist == IST_DOT_CHAR { 1 } else { 0 }
		c := if re.prog[pc1+2].ist == IST_DOT_CHAR { 1 } else { 0 }
		d := if re.prog[pc1+3].ist == IST_DOT_CHAR { 1 } else { 0 }
		e := if re.prog[pc1+4].ist == IST_DOT_CHAR { 1 } else { 0 }
		f := if re.prog[pc1+5].ist == IST_DOT_CHAR { 1 } else { 0 }
		g := if re.prog[pc1+6].ist == IST_DOT_CHAR { 1 } else { 0 }
		h := if re.prog[pc1+7].ist == IST_DOT_CHAR { 1 } else { 0 }
	}
	for ; pc1 < pc; pc1++ {
		if re.prog[pc1].ist == IST_DOT_CHAR {
			tmp_count++
		}
	}
```

**Comment by penguindark** [comment]
> Maybe current compilers are very smart and will always discover it even if it's not the standard counted loop, but to increase the probability, I'd do change it .

I had the same idea and tested it, I'm using gcc as backend , and today compiler are very smart in unroll the loops.
After few tests I noticed no difference using `-O3` optimization and decided to let the loop and let the compiler unroll it, the only benefit is that the code is more didactic.
Other note it is that the 95% of the strings will be ascii 8 bit, thus the loop work for only on time and this fact create and advantage for the loop implementation.
This is very interesting aspect of modern compiler and CPU.
Nowadays broke the CPU pipeline it is very difficult with all the parallel units dedicated to the loop execution.
But you guesses are all right ;)

**Comment by penguindark** [comment]
> could be optimized by matching e.g. 4 or better 8 characters conditions in one iteration 

IST_DOT_CHAR  is a ugly beast :(
you can check only one for time and after you **must** check the following tokens,
otherwise expression like `\s(.*)abab.*pe` will not work properly.

**Comment by penguindark** [comment]
the idea how to unroll the IST_DOT_CHAR, the methodology I mean is very good :) I like it! Very smart! :+1:

**Comment by penguindark** [comment]
> In parse_char_class() I think all references to BSLS_VALIDATOR_ARRAY[c] could be changed to x - is it so?

Can you elaborate, I'm not sure that I understood well your idea, I apologize for this.

**Comment by dumblob** [comment]
>After few tests I noticed no difference using `-O3` optimization and decided to let the loop and let the compiler unroll it,

Just a note regarding `-O3` - be very careful as `-O3` is really not secure and therefore not used in production (there is even some issue in the *V* issue tracker and *V* uses therefore `-O2` if I'm not mistaken). So, I would rather test all unrolling etc. just with `-O2` :wink:.

**Comment by penguindark** [comment]
I test with standard, O2, O3 and always do memory check ;) I know that `gcc -O3` have a lot of defect but not soo much defect after all.
I use GCC from, I dont even remember, but it can put very sneaky tricks on your code and make you mad trying to understand why your code run in normal mode and obtain different result with O3... I know well, on my skin, I spent nights chasing these bugs -_-

And from what you write you too! :wink:

**Comment by dumblob** [comment]
>Can you elaborate, I'm not sure that I understood well your idea, I apologize for this.

Sure. I thought the following code has a minor typo. Namely there is the `c, x` (index, value), but then the array is unnecessarily accessed even though the value stored in `x` is already available :wink:.
```v
			for c,x in BSLS_VALIDATOR_ARRAY {
				if x.ch == ch {
					//C.printf("CC bsls found \\%c.\n",ch)
					re.cc[tmp_index].cc_type   = CC_BSLS
					re.cc[tmp_index].ch0       = BSLS_VALIDATOR_ARRAY[c].ch
					re.cc[tmp_index].ch1       = BSLS_VALIDATOR_ARRAY[c].ch
					re.cc[tmp_index].validator = BSLS_VALIDATOR_ARRAY[c].validator
					i += char_len
					tmp_index++
					status = .in_char
					break
				}
			}
```
would become
```v
			for c,x in BSLS_VALIDATOR_ARRAY {
				if x.ch == ch {
					//C.printf("CC bsls found \\%c.\n",ch)
					re.cc[tmp_index].cc_type   = CC_BSLS
					re.cc[tmp_index].ch0       = x.ch
					re.cc[tmp_index].ch1       = x.ch
					re.cc[tmp_index].validator = x.validator
					i += char_len
					tmp_index++
					status = .in_char
					break
				}
			}
```

**Comment by penguindark** [comment]
Ahh ok, sure you are right, I'll check, tnx :)

**Comment by dumblob** [comment]
Ok, I've read the whole code and I really can't find any good places for optimization without running profiling tool.

There are just two more thoughts:

1. I'd change `pub fn (re mut RE) match_base(in_txt byteptr, in_txt_len int ) (int,int) {` to a signature which doesn't explicitly use any pointers. It's because if in the future *V* will use `restrict` (for compilation into `C`), then code using explicit pointers (in contrast to the implicit pointer `mut`) will be more prone to not gain any speedup. So, I'd avoid explicit pointers in *V* as much as possible. Actually in this case one could use plain `string` instead of `byteptr, int`.

2. I'd maybe try to simplify things for the compiler and instead of:
```v
			if re.state_stack_index >=0 && re.state_stack[re.state_stack_index].pc >= 0 {
				i = re.state_stack[re.state_stack_index].i
				pc = re.state_stack[re.state_stack_index].pc
				state.match_index =	re.state_stack[re.state_stack_index].mi
				group_index = re.state_stack[re.state_stack_index].group_stack_index

				m_state = .ist_load
				continue
			}
```
I'd write something like:
```v
			if re.state_stack_index >=0 {
				// saves some dereferences in case of a bad compiler unable to analyze, that there were no changes to the pointer in between the read-outs
				a := re.state_stack[re.state_stack_index]
				if a.pc >= 0 {
					i = a.i
					pc = a.pc
					state.match_index = a.mi
					group_index = a.group_stack_index

					m_state = .ist_load
					continue
				}
			}
```
(this pattern appears also on few other places :wink:)

Otherwise we would need to talk about added caching etc. But mainly doing a proper profiling (flame graphs are super cool) with `-O2` to get an overview of what is slowing us down.

Thanks for your very nice work overall. I'm pleased to read such nice code with a lot of meaningful functionality squeezed in just a few lines!

**Comment by penguindark** [comment]
Tnx a lot for your help, your ideas and suggestion are precious to me :+1:  tnx.

> I'd change `pub fn (re mut RE)...`

I'd like too, 
that code was written to avoid a lot of memory passages using V strings, for sure the code could be more pleasant buf for sure more slower (not that the actual code si so fast, we are near six time slower then python regex!! :( )
I'm working to have in the V2 compiler a "read only" copy of a string or substring to use for these cases.
The first version I wrote used only V strings and slice operation on them, but was awful slower :( 
after that deludent result, I rewritten all the code.. sigh..

> I'd write something like: ...

Yeah you are right, this is one of the optimization I was thinking to do, you are right.
I need to check first how is the asm code for the two solutions, I'll keep you updated on this :)
I've only a doubt about cache mapping for the structs, but we will see.
